contracts/Graph.sol

```diff
@@ -1,103 +1,177 @@
+    function addMultipleEdges(address[] calldata recipients, uint256[] calldata weightsToAdd) public {
+        uint256 senderId = addressToId[msg.sender];
+        uint256 length = recipients.length;
+
+        for (uint256 i = 0; i < length; i++) {
+            uint256 recipientId = addressToId[recipients[i]];
             sourceIds.push(senderId);
             targetIds.push(recipientId);
             weights.push(weightsToAdd[i]);
         }
 
     }
 
-
-    uint public constant DECIMALS = 1000000000000000000;
-
-    function getPageRank(address _address) public view returns (uint) {
+    function getPageRank(address _address) public view returns (uint256) {
         return 1;
     }
 
-    function getMyPageRank() public view returns (uint) {
+    function getTotalPageRank(address _address) public view returns (uint256) {
+        return 100;
+    }
+
+    function getMyPageRank() public view returns (uint256) {
         return getPageRank(msg.sender);
     }
 
-    mapping(address => uint) public addressToWithdrawalAmount;
+    mapping(address => uint256) public addressToWithdrawalAmount;
 
     function fund() public payable {
         totalBudget += msg.value;
         historicalBudget += msg.value;
+        emit depositEvent(msg.sender, msg.value);
     }
 
-    uint public totalBudget;
-
-    uint public monthlyBudget;
-    uint public historicalBudget; 
-    function withdraw(uint amount) public { // Accounts are entitled to funds proportional to their pagerank. They should not be able to withdraw more than their budget: min(M*p/sum(p), budget)
-        uint amount = addressToWithdrawalAmount[msg.sender];     
-
-
-        uint available = monthlyBudget*getMyPageRank()/DECIMALS - addressToWithdrawalAmount[msg.sender];
-        
-        require(amount <= available, "Insufficient funds");
-        addressToWithdrawalAmount[msg.sender] += amount;
-        totalBudget -= amount;
+    function withdraw(uint256 _amount) public  noReentrancy{ // Accounts are entitled to funds proportional to their pagerank. They should not be able to withdraw more than their budget: min(M*p/sum(p), budget)
+        require(organisations[msg.sender] == true, "You are not an organinsation");
+        require(withdrawStatus == State.OPENED, "Withdraw is not Opened");
+        require(monthlyBudget > _amount, "Insufficent for Withdrawal");
+        require(mywithdrawstatus[msg.sender], "Exceeded you withdraw limit");
+        /** @dev 
+         * Calculate the expected Withdraw of a user or organisation
+        */
+        uint256 expectedWithdraw = uint256(monthlyBudget).mul(uint256(getPageRank(msg.sender))).div(uint256(getTotalPageRank(msg.sender)));
+       
+        require(expectedWithdraw > monthlyWithdraw[msg.sender]);
+                
+        monthlyWithdraw[msg.sender] += _amount;
+        userTotalWithdraw[msg.sender] += _amount;
+        totalBudget -= _amount;
+        monthlyBudget -= _amount;
```

Comment:
I don't think the monthlyBudget should be reduced. This will mean people who withdraw later on will get an unfairly small amount. I.e. if two people each have 10% of the total pagerank, the second person will only get 0.9*0.1=0.09 instead of both getting 0.1